> Hi @orionsBeltWest ,
> 
> We udated the Dockerfile.hpu (https://github.com/BlueLabelLabs/whisper_fork/pull/2/files#diff-8e8c9c8166fa53e2eb5a4fa3ea94131bf7e45e608f8161bbcee88c735fe171fb) following vllm (https://github.com/vllm-project/vllm/blob/main/Dockerfile.hpu) and also made an update in section Intel® Gaudi® hpu usage of README.md - https://github.com/BlueLabelLabs/whisper_fork/pull/2/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R147. Added an info about optional mapping using docker run command and also added an info about running the HPU tests inside the docker container - https://github.com/BlueLabelLabs/whisper_fork/pull/2/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R178



No I am not able to run this build due to RuntimeError: synStatus=26 [Generic failure] Device acquire failed.

I ran it both in baremetal and k8 same issue. 


root@98ff692c3ac2:/workspace/whisper# python3 -m whisper.transcribe keynote.wav --model turbo --device hpu
/usr/lib/python3.10/runpy.py:126: RuntimeWarning: 'whisper.transcribe' found in sys.modules after import of package 'whisper', but prior to execution of 'whisper.transcribe'; this may result in unpredictable behaviour
  warn(RuntimeWarning(msg))
/usr/lib/python3.10/inspect.py:288: FutureWarning: `torch.distributed.reduce_op` is deprecated, please use `torch.distributed.ReduceOp` instead
  return isinstance(object, types.FunctionType)
Loading Habana modules from /usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/lib
/workspace/whisper/whisper/__init__.py:154: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  checkpoint = torch.load(fp, map_location="cpu")
Loading Habana modules from /usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/lib
/home/jenkins/workspace/cdsoftwarebuilder/create-binaries-from-sw-sources---bp-dt/repos/hcl/src/platform/gaudi_common/hcl_device_control_factory.cpp::84(initDevice): The condition [ g_ibv.init(deviceConfig) == hcclSuccess ] failed. ibv initialization failed 
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/workspace/whisper/whisper/transcribe.py", line 625, in <module>
    cli()
  File "/workspace/whisper/whisper/transcribe.py", line 597, in cli
    model = load_model(model_name, device=device, download_root=model_dir)
  File "/workspace/whisper/whisper/__init__.py", line 173, in load_model
    model = wrap_in_hpu_graph(model)
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpu/graphs.py", line 720, in wrap_in_hpu_graph
    stream = htorch.hpu.default_stream()
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpu/streams.py", line 284, in default_stream
    streamdata = _hpu_C._hpu_getDefaultStream(_get_device_index(device, optional=True))
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpu/_utils.py", line 42, in _get_device_index
    device_idx = hpu.current_device()
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpu/__init__.py", line 132, in current_device
    init()
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpu/__init__.py", line 71, in init
    _hpu_C.init()
RuntimeError: synStatus=26 [Generic failure] Device acquire failed.
Exception ignored in atexit callback: <built-in method cleanup of PyCapsule object at 0x7efb62607c30>
RuntimeError: Habana device not initialized
Exception ignored in atexit callback: <built-in method cleanup of PyCapsule object at 0x7efb62607c30>
RuntimeError: Habana device not initialized
root@98ff692c3ac2:/workspace/whisper# hl-smi 
+-----------------------------------------------------------------------------+
| HL-SMI Version:                              hl-1.18.0-fw-53.1.1.1          |
| Driver Version:                                     1.18.0-ee698fb          |
|-------------------------------+----------------------+----------------------+
| AIP  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | AIP-Util  Compute M. |
|===============================+======================+======================|
|   0  HL-225              N/A  | 0000:0e:00.0     N/A |                   0  |
| N/A   28C   N/A    72W / 600W |    768MiB / 98304MiB |     0%           N/A |
|-------------------------------+----------------------+----------------------+
| Compute Processes:                                               AIP Memory |
|  AIP       PID   Type   Process name                             Usage      |
|=============================================================================|
|   0        N/A   N/A    N/A                                      N/A        |
+=============================================================================+
root@98ff692c3ac2:/workspace/whisper#

--- a/tests/test_transcribe.py
+++ b/tests/test_transcribe.py
@@ -51,7 +51,7 @@ def test_transcribe_hpu(model_name: str):
 
     language = "en" if model_name.endswith(".en") else None
     result = model.transcribe(
-        audio_path, language=language, temperature=0.0
+        audio_path, language=language, temperature=0.0, word_timestamps=True
     )
     assert result["language"] == "en"
     assert result["text"] == "".join([s["text"] for s in result["segments"]])

--- a/whisper/timing.py
+++ b/whisper/timing.py
@@ -10,6 +10,7 @@ import torch
 import torch.nn.functional as F
 
 from .audio import HOP_LENGTH, SAMPLE_RATE, TOKENS_PER_SECOND
+from .hpu_utils import is_hpu_device
 from .tokenizer import Tokenizer
 
 if TYPE_CHECKING:
@@ -251,7 +252,21 @@ def find_alignment(
         hook.remove()
 
     # heads * tokens * frames
-    weights = torch.stack([QKs[_l][_h] for _l, _h in model.alignment_heads.indices().T])
+    # Adjust alignment head indices for HPU
+    weights = []
+    if is_hpu_device(model.device):
+        # Handle dense layout for HPU
+        alignment_heads_dense = model.alignment_heads.to_dense() if model.alignment_heads.is_sparse else model.alignment_heads
+        indices = alignment_heads_dense.nonzero(as_tuple=True)
+        for _l, _h in zip(*indices):
+            weights.append(QKs[_l][_h])
+    else:
+        # Default behavior for non-HPU devices
+        for _l, _h in model.alignment_heads.indices().T:
+            weights.append(QKs[_l][_h])
+
+    # Stack the weights
+    weights = torch.stack(weights)
     weights = weights[:, :, : num_frames // 2]
     weights = (weights * qk_scale).softmax(dim=-1)
     std, mean = torch.std_mean(weights, dim=-2, keepdim=True, unbiased=False)

